### PR TITLE
Gemini Support

### PR DESCRIPTION
**Pull request by valaises**
https://refact.fibery.io/Software_Development/New-document-428

### CODE REVIEW

File: refact_webgui/webgui/selfhost_static.py
Comment (valaises) [inline refact_webgui/webgui/selfhost_static.py:1]: oops, gotta remove

File: refact_webgui/webgui/selfhost_fastapi_completions.py [hunk 1]
```diff
@@ -337,7 +339,10 @@ async def _passthrough_tokenizer(self, model_path: str) -> str:
         try:
             async with aiohttp.ClientSession() as session:
                 tokenizer_url = f"https://huggingface.co/{model_path}/resolve/main/tokenizer.json"
-                async with session.get(tokenizer_url) as resp:
+                headers = {}
+                if hf_token := huggingface_hub_token():
+                    headers["Authorization"] = f"Bearer {hf_token}"
+                async with session.get(tokenizer_url, headers=headers) as resp:
```
Comment (mitya52) [inline refact_webgui/webgui/selfhost_fastapi_completions.py:345]: this is not a good practice for 3rdpartz models

File: setup.py [hunk 2]
```diff
@@ -45,7 +45,7 @@ class PyPackage:
                   "bitsandbytes", "safetensors", "peft", "triton",
                   "torchinfo", "mpi4py", "deepspeed>=0.15.3",
                   "sentence-transformers", "huggingface-hub>=0.26.2",
-                  "aiohttp", "setproctitle"],
+                  "aiohttp", "setproctitle", "google-auth>=2.37.0"],
```
Comment (mitya52) [inline setup.py:48]: google-auth? for what purpose?

File: refact_known_models/passthrough.py [hunk 3]
```diff
@@ -209,5 +209,55 @@
...
+    # gemini and gemma bear the same tokenizer
+    # according to https://medium.com/google-cloud/a-gemini-and-gemma-tokenizer-in-java-e18831ac9677
+    # downloadable tokenizer.json does not exist for gemini, proposed solution to use vertexai lib in python uses web requests
+    # for pricing consult: https://ai.google.dev/pricing
+    # pricing below is assumed for <= 128_000 context is used
+
+    "gemini-2.0-flash-exp": {
+        "backend": "litellm",
+        "provider": "gemini",
+        "tokenizer_path": "google/gemma-7b",
```
Comment (mitya52) [inline refact_known_models/passthrough.py:223]: https://huggingface.co/Xenova/gemma2-tokenizer/tree/main

**Review by mitya52 — CHANGES_REQUESTED** [review]

File: setup.py (hunk 2 above)
Comment (valaises) [reply setup.py:48]: litellm requires it to refer to google api

File: refact_webgui/webgui/selfhost_fastapi_completions.py [hunk 4]
```diff
@@ -13,6 +13,7 @@
 from fastapi import APIRouter, HTTPException, Query, Header
 from fastapi.responses import Response, StreamingResponse
 
+from refact_utils.huggingface.utils import huggingface_hub_token
```
Comment (mitya52) [inline refact_webgui/webgui/selfhost_fastapi_completions.py:16]: remove it please

**Review by mitya52 — CHANGES_REQUESTED** [review]

**Review by mitya52 — APPROVED** [review]

**Pull request merged**